platform/MCU/STM32F2xx/STM32_USB_Device_Driver/src/usbd_cdc_core.c

```diff
@@ -626,44 +693,65 @@ static uint8_t  usbd_cdc_EP0_RxReady (void  *pdev)
   */
 static uint8_t  usbd_cdc_DataIn (void *pdev, uint8_t epnum)
 {
-  uint16_t USB_Tx_ptr;
-  uint16_t USB_Tx_length;
-
-  if (USB_Tx_State == 1)
-  {
-    if (APP_Rx_length == 0)
-    {
-      USB_Tx_State = 0;
-    }
-    else
-    {
-      if (APP_Rx_length > CDC_DATA_IN_PACKET_SIZE){
-        USB_Tx_ptr = APP_Rx_ptr_out;
-        USB_Tx_length = CDC_DATA_IN_PACKET_SIZE;
+  uint32_t USB_Tx_length;
 
-        APP_Rx_ptr_out += CDC_DATA_IN_PACKET_SIZE;
-        APP_Rx_length -= CDC_DATA_IN_PACKET_SIZE;
-      }
-      else
-      {
-        USB_Tx_ptr = APP_Rx_ptr_out;
-        USB_Tx_length = APP_Rx_length;
+  if (!USB_Tx_State)
+    return USBD_OK;
 
-        APP_Rx_ptr_out += APP_Rx_length;
-        APP_Rx_length = 0;
-      }
+  USB_Tx_length = ring_data_contig(USB_TX_BUFFER_SIZE, USB_Tx_Buffer_head, USB_Tx_Buffer_tail);
 
-      /* Prepare the available data buffer to be sent on IN endpoint */
-      DCD_EP_Tx (pdev,
-                 CDC_IN_EP,
-                 (uint8_t*)&APP_Rx_Buffer[USB_Tx_ptr],
-                 USB_Tx_length);
-    }
+  if (USB_Tx_length) {
+    USB_Tx_length = MIN(USB_Tx_length, CDC_DATA_IN_PACKET_SIZE);
+  } else if (usbd_Last_Tx_Packet_size(pdev, epnum) != CDC_DATA_IN_PACKET_SIZE) {
+    USB_Tx_State = 0;
+    return USBD_OK;
   }
 
+  /* Prepare the available data buffer to be sent on IN endpoint */
+  DCD_EP_Tx (pdev,
+             CDC_IN_EP,
+             (uint8_t*)&USB_Tx_Buffer[USB_Tx_Buffer_tail],
+             USB_Tx_length);
+
+  USB_Tx_Buffer_tail = ring_wrap(USB_TX_BUFFER_SIZE, USB_Tx_Buffer_tail + USB_Tx_length);
+  USB_Tx_Buffer_length += USB_Tx_length;
   return USBD_OK;
 }
 
+static inline int usbd_cdc_Start_Rx(void *pdev)
+{
+  //
+  uint32_t USB_Rx_length;
+  if (USB_Rx_Buffer_head >= USB_Rx_Buffer_tail)
+    USB_Rx_Buffer_length = USB_RX_BUFFER_SIZE;
```

Comment:
Can you add code comments please to clarify the need for a buffer length in addition to the head/tail pointers?